### ISSUE TITLE
Client searching and "client configuration validation" runs multiple times, and produces 10 SQL queries via EF each run

### ISSUE THREAD

**Issue by felinepc** [DuendeSoftware/IdentityServer#106]
**Which exact version are you using?**

5.0.3

**Describe the bug**

I first noticed that local API calls against my controllers on IdentityServer protected by `AddLocalApiAuthentication` are slowish. So I reviewed the IdentityServer logs, also turned on EF loggin, and found that each time token validation occurs, it goes through:

> 10 SQL queries against the client store (due to the way EF queries in `ClientStore.cs` was written)
>  Duende.IdentityServer.EntityFramework.Stores.ClientStore
> `MyClient` found in database: True
> client configuration validation for client `MyClient` succeeded.
> 
> Repeat 10 SQL queries against the client store
> Duende.IdentityServer.EntityFramework.Stores.ClientStore
> `MyClient` found in database: True
> client configuration validation for client `MyClient` succeeded.
> 
> Duende.IdentityServer.Validation.TokenValidator
> Calling into custom token validator: Duende.IdentityServer.Validation.DefaultCustomTokenValidator
> Duende.IdentityServer.Validation.TokenValidator
> Token validation success

My first question was why the client validation has to be ran for local API authentication, let alone twice? Is it possible to have `IdentityServer` simply validate the access token against its own signature (like how a third party RP would do) without making DB calls?

Secondly, why `ClientStore.cs` was written to execute the 9 extra queries separately instead of using one eager load?

Then I reviewed the log for a normal OIDC Authorization Code flow initiated from my app. This time, I counted a total of 6 x client configuration validation runs (so 60 SQL queries for them). 

I can post full log if needed (but they're pretty straight forward as described here).

Is this normal or something could be wrong?

**Comment by leastprivilege** [comment]
Have you turned on caching for the client store?

https://docs.duendesoftware.com/identityserver/v5/data/configuration/#caching-configuration-data

**Comment by felinepc** [comment]
Ahh, missed that part of the documentation when I configured EF operation stores. So I added this:
```
builder.AddInMemoryCaching().AddClientStoreCache<ClientStore>()
                .AddConfigurationStoreCache()
                .AddResourceStoreCache<ResourceStore>()
                .AddCorsPolicyCache<CorsPolicyService>();
```
and slowness and queries did go away. It doesn't even produce the client configuration validation logs any more. I assume this is using all default IdentityServer in-memory caching.

I'm still curious on why it would perform those validations against database so many times without caching (I guess it's some low level internal implementations/verifications for security?), and why the default EF `ClientStore` was written to do 10 individual queries instead of one Include/join.
